src/Learning/KWDataPreparation/KWClassStats.cpp

```diff
@@ -1128,34 +1142,40 @@ void KWClassStats::WriteJSONFields(JSONFile* fJSON)
 	fJSON->WriteKeyString("dictionary", GetClass()->GetName());
 
 	// Nombres d'attributs par type
-	fJSON->BeginKeyObject("variables");
-	fJSON->BeginKeyArray("types");
-	for (nType = 0; nType < KWType::None; nType++)
+	if (IsStatsComputed())
 	{
-		if (KWType::IsData(nType))
+		fJSON->BeginKeyObject("variables");
+		fJSON->BeginKeyArray("types");
+		for (nType = 0; nType < KWType::None; nType++)
 		{
-			nAttributeNumber = GetUsedAttributeNumberForType(nType);
-			if (nAttributeNumber > 0)
-				fJSON->WriteString(KWType::ToString(nType));
+			if (KWType::IsData(nType))
+			{
+				nAttributeNumber = GetUsedAttributeNumberForType(nType);
+				if (nAttributeNumber > 0)
+					fJSON->WriteString(KWType::ToString(nType));
+			}
 		}
-	}
-	fJSON->EndArray();
-	fJSON->BeginKeyArray("numbers");
-	for (nType = 0; nType < KWType::None; nType++)
-	{
-		if (KWType::IsData(nType))
+		fJSON->EndArray();
+		fJSON->BeginKeyArray("numbers");
+		for (nType = 0; nType < KWType::None; nType++)
```

Comment:
I would try to factor out this loop so that we don't need to run it twice, thusly calling `GetUsedAttributeNumberForType` twice as many times. Perhaps, running the loop only once and populating two vectors with the `nType`s and `nAttributeNumber`s, respectively, then iterating on each of the vectors for writing their values in the appropriate JSON fields?

Reply:
C'est un peu de l'over-killing
- il n'y a ici aucun d'enjeu d'optimisation
- c'est un simple reporting, et la deuxième boucle obtenue par copier-coller de la première (et juste en dessous) est particulièrement simple à lire

Reply:
Ok, it seems more readable and straightforward as it is, indeed.